File: STGProject/Source/Mdl/STG/MyShip/TypeA/Laser.cpp
#include "PCH.h"
#include "Laser.h"
#include "Resource.h"

#include "Ctrl/STG/Actors.h"
#include "Mdl/STG/Actor/Particle/Creator.h"
#include "View/STG/DrawPriority.h"

using namespace Game;
using namespace Mdl::STG;
using namespace MyShip::TypeA;
using namespace Util::STG;
using namespace Selene;

namespace
{
	static const int RADIUS = 6;
	static const Vector2DF LASER_SIZE( 12, 32 );
	static const float ANGLE = 270.0f;
	static const float SPEED = 22.0f;
	static const float ALPHA = 0.8f;
	static const float DAMAGE = 10.0f;
}


// コンストラクタ
Laser::Laser( Base::PMode pMode, const Vector2DF &pos, 
			 Util::Sprite::PTexture pLaserTexture, Util::Sprite::PTexture pParticleTexture )
: Base( pMode, pos, RADIUS, DAMAGE )
, mpParticleTexture( pParticleTexture )
, mAuxs( pMode )
{
	assert( pLaserTexture );
	assert( pParticleTexture );

	mLocator.GetPosition() = pos;
	mLocator.GetSpeed().SetUnitVector( ANGLE ) *= SPEED;

	Base::SetDamageRateDec( 0.06f );
	Base::SetDamageRateMin( 0.5f );

	mDrawParam.SetTexture( pLaserTexture );
	mDrawParam.SetSrc( Resource::LASER.GetRect() );
	mDrawParam.SetDst( mLocator.GetPosition().MakeRect( LASER_SIZE ) );
	mDrawParam.SetPriority( View::STG::PRI_MYSHOT );
	mDrawParam.SetRotDegree( ANGLE - 270.0f );
	mDrawParam.SetColor( ColorF( 1.0f, 1.0f, 1.0f, ALPHA ) );
}


// 位置の取得
Vector2DF Laser::GetPosition() const
{
	return mLocator.GetPosition();
}

// 外力の適用
Vector2DF Laser::ApplyExternalForce( const Vector2DF &vec )
{
	return mLocator.ApplyExternalForce( vec );
}


void Laser::OnUpdate()
{
	mLocator.Update();

	mDrawParam.SetDst( mLocator.GetPosition().MakeRect( LASER_SIZE ) );
	mDrawParam.SetRotDegree( mLocator.GetSpeed().GetAngle() - 270.0f );
}

void Laser::OnDraw() const
{
	Util::Sprite::Manager::Draw( mDrawParam );
}

void Laser::OnDestruct()
{
	using namespace Mdl::STG::Actor;

	Particle::Parameter param;
	{
		Util::Sprite::DrawParameter dParam;
		dParam.SetTexture( mpParticleTexture );
		dParam.SetSrc( Resource::BLUE_CIRCLE.GetRect() );
		dParam.SetDst( mLocator.GetPosition().MakeRect( 12, 12 ) );
		dParam.SetPriority( View::STG::PRI_MYSHOT );
		dParam.SetAlphaBlend( Util::Sprite::AB_ADD );
		param.SetDrawParameter( dParam );
	}
	param.SetPosition( mLocator.GetPosition() + Vector2DF( 0, -6 ) );
	param.SetAngleDomain( ANGLE+180 - 20, ANGLE+180 + 20 );
	param.SetSpeedDomain( 8.0f, 12.0f );
	param.SetZoomRateDomain( 0.8f, 1.0f );
	param.SetZoomDecRateDomain( 0.06f, 0.08f );

	Particle::Creator creator( Base::GetMode(), param );
	creator.Create( 30 );
}

void Laser::OnDestruct( float hp )
{
	mAuxs.CreateFadeOut( mDrawParam );
}

void Laser::OnErase()
{
	mAuxs.CreateFadeOut( mDrawParam );
}

File: Script/Python/src/STG/MyShip/TestShip/Laser.py
'''
Created on 2010/11/10

@author: t
'''

import Std
import STG.Ctrl as Ctrl
import STG.Auxs as Auxs
from STG.MyShot.BaseMyShot import Base
import STG.Resource.Common as CommonResource

class Laser(Base):
	RADIUS = 3
	SIZE = Std.Vector2DF(40, RADIUS*2)
	SPEED = 30
	BASE_DAMAGE = 4
	DAMAGE_RATE_DEC = 0.06
	DAMAGE_RATE_MIN = 0.5
	
	def __init__(self, position, angle):
		super().__init__(position, self.RADIUS, self.BASE_DAMAGE)
		
		self.locator = Std.Locator.LinearF(position)
		self.locator.speed.setUnitVector(angle)
		self.locator.speed *= self.SPEED
		
		self.damageRateDec = self.DAMAGE_RATE_DEC
		self.damageRateMin = self.DAMAGE_RATE_MIN
		
		self.drawParam = Std.Sprite.DrawParam()
		self.drawParam.texture = CommonResource.get("bulletGreen")
		self.drawParam.src = Std.RectF(0, 0, 64, 64)
		self.drawParam.dst = self.locator.position.makeRect(self.SIZE, True)
		self.drawParam.priority = Ctrl.DrawPriority.myShot
		self.drawParam.rotDegree = angle
		self.drawParam.color = Std.ColorF(1, 1, 1, 0.8)
	
	@property
	def position(self):
		return self.locator.position
	def applyExternalForce(self, vec):
		return self.locator.applyExternalForce(vec)
	
	def onUpdate(self):
		self.locator.update()
		self.drawParam.dst = self.locator.position.makeRect(self.SIZE, True)
		self.drawParam.rotDegree = self.locator.speed.getAngle()
	def onDraw(self):
		Std.Sprite.draw(self.drawParam)
	
	def onErase(self):
		Auxs.createFadeOut(self.drawParam)
	def onDestructed(self):
		drawParam = Std.Sprite.DrawParam()
		drawParam.texture = CommonResource.get("bulletGreen")
		drawParam.src = Std.RectF(0, 0, 64, 64)
		drawParam.dst = self.locator.position.makeRect(12, 12, True)
		drawParam.priority = Ctrl.DrawPriority.myShot
		drawParam.alphaBlend = Std.Sprite.AlphaBlend.Add
		param = Auxs.Particle.Param()
		param.drawParameter = drawParam
		
		angle = self.locator.speed.getAngle()
		direction = Std.Vector2DF()
		direction.setUnitVector(angle)
		param.setPosition(self.locator.position + direction*10)
		param.setAngleRange(angle+180 - 20, angle+180 + 20)
		param.setSpeedRange(8, 12)
		param.setZoomRateRange(0.8, 1)
		param.setZoomDecRateRange(0.06, 0.08)
		
		creator = Auxs.Particle.Creator(param)
		creator.create(2)
	def onDamaged(self, damage):
		pass
	def onDestructedByDamage(self, damage):
		self.onErase()